lib/cache/reliability_manager.js

```diff
@@ -0,0 +1,164 @@
+const cluster = require('cluster');
+const path = require('path');
+const helpers = require('../helpers');
+const cm = require('../cluster_messages');
+const consts = require('../constants');
+const crypto = require('crypto');
+const _ = require('lodash');
+
+const kDbVersionReliability = 'version_reliability_manager_versions';
+const kUnreliableRootDir = '.unreliable';
+const defaultOptions = {
+    reliabilityThreshold: 0,
+    multiClient: false
+};
+
+class ReliabilityManager {
+    constructor(db, cachePath, options) {
+        this._db = db;
+        this._cachePath = cachePath;
+        this._options = options || {};
+        _.defaults(this._options, defaultOptions);
+
+        if(this._db) {
+            this._db_versionReliability = db.getCollection(kDbVersionReliability);
+            if (this._db_versionReliability == null) {
+                this._db_versionReliability = db.addCollection(kDbVersionReliability, {
+                    unique: ["versionId"],
+                    indices: ["guid", "hash"]
+                });
+            }
+        }
+
+        cm.listenFor("_updateReliabilityFactorForVersion", (data) => {
+            return this._updateReliabilityFactorForVersion(data);
+        });
+    }
+
+    /**
+     *
+     * @param {Object} params
+     * @returns {Promise<Object>}
+     * @private
+     */
+    async _updateReliabilityFactorForVersion(params) {
+        if(cluster.isWorker) {
+            return cm.send('_updateReliabilityFactorForVersion', params);
+        }
+
+        const entry = this.getEntry(params.guidStr, params.hashStr, true);
+        if(!entry.versionHash) {
+            entry.versionHash = params.versionHashStr;
+            entry.clientId = params.clientId;
+        }
+
+        if(entry.state !== ReliabilityManager.reliabilityStates.Pending) {
+            if(entry.state === ReliabilityManager.reliabilityStates.ReliableNew) {
+                entry.state = ReliabilityManager.reliabilityStates.Reliable;
+                this._db_versionReliability.update(entry);
+            }
+
+            return entry;
+        }
+
+        if(this._options.multiClient && params.clientId === entry.clientId) {
+            helpers.log(consts.LOG_DBG, `Ignoring duplicate transaction for GUID: ${params.guidStr} Hash: ${params.hashStr} from previous client (multiClient = true)`);
+            return entry;
+        }
+
+        entry.clientId = params.clientId;
```

Comment:
With two clients "A" and "B" if transactions come in the order A->B->A->B, it would end up with a reliabilityFactor of 4, but transactions in the order A->A->B->B, would have a reliabilityFactor of 2. Seems a little odd that the same transition set can end up with differing reliabilityFactors if the timing changes a bit. Is that intended?

Reply:
Yes, this is intended - I don't want to actually track unique clients. In practice, I believe 99% of the time the default threshold of 2 will be used, so we would hardly gain from the added complexity of keeping track of all unique clients.